File: programmers/지게차와 크레인.py
from collections import deque


# 뺀 건 ' ', 이번 텀에 뺀 건 'b',
def bfs(graph, v, qx, qy, dx, dy, n, m):
    visited = [[False] * m for _ in range(n)]
    queue = deque()
    queue.append((qx, qy))
    visited[qx][qy] = True
    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < m and not visited[nx][ny]:
                if graph[nx][ny] == v:
                    graph[nx][ny] = 'b'
                if graph[nx][ny] == ' ':
                    queue.append((nx, ny))
                visited[nx][ny] = True


def bToa(graph):
    for i, gr in enumerate(graph):
        for j, g in enumerate(gr):
            if g == 'b':
                graph[i][j] = ' '


def solution(storage, requests):
    graph = [list(' ' * (len(storage[0]) + 2))]
    for s in storage:
        g = [' ']
        for ss in s:
            g.append(ss)
        g.append(' ')
        graph.append(g)
    graph.append(list(' ' * (len(storage[0]) + 2)))

    dx = [-1, 0, 1, 0]
    dy = [0, -1, 0, 1]
    n = len(graph)
    m = len(graph[0])
    print(n, m)

    for r in requests:
        if len(r) == 1:
            bfs(graph, r, 0, 0, dx, dy, n, m)
            bfs(graph, r, n - 1, 0, dx, dy, n, m)
            bfs(graph, r, 0, m - 1, dx, dy, n, m)
            bfs(graph, r, n - 1, m - 1, dx, dy, n, m)
            bToa(graph)
        else:
            for i, gr in enumerate(graph):
                for j, g in enumerate(gr):
                    if g == r[0]:
                        graph[i][j] = ' '

    cnt = 0
    for gr in graph:
        for g in gr:
            if g != ' ':
                cnt += 1
    return cnt

File: programmers/지게차와 크레인.js
// 테두리가 비어있는 그래프 생성 -> 빈 공간만 지게차가 이동 가능
// 알파벳 하나 - 접근 가능한 컨테이너 (전부)
// - 상하좌우 체크 후 빈 공간 or 테두리 인 경우만 제거
// 알파벳 두개 - 모든 컨테이너

function solution(storage, requests) {
    const n = storage.length
    const m = storage[0].length
    const dx = [-1, 1, 0, 0]
    const dy = [0, 0, -1, 1]
    
    const graph = Array.from({ length: n + 2 }, () => Array(m + 2).fill(''))
    for (let i=1; i<n+1; i++) {
        for (let j=1; j<m+1; j++) {
            graph[i][j] = storage[i - 1][j - 1]
        }
    }
    
    // 지게차 출고 시
    const bfs = (req) => {
        const visited = Array.from({ length: n + 2 }, () => Array(m + 2).fill(false))
        const queue = [[0, 0]]
        visited[0][0] = true
        let head = 0
        
        const arr = []
        
        while (head < queue.length) {
            const [x, y] = queue[head++]
            for (let i=0; i<4; i++) {
                const nx = x + dx[i]
                const ny = y + dy[i]
                if (nx >= 0 && nx < m+2 && ny >= 0 && ny < n+2 && !visited[ny][nx]) {
                    if (graph[ny][nx] === '') queue.push([nx, ny])
                    else if (graph[ny][nx] === req) arr.push([nx, ny])
                    visited[ny][nx] = true
                }
            }
        }
        
        for (const [x, y] of arr) {
            graph[y][x] = ''
        }
    }
    
    const two = (req) => {
        for (let i=1; i<n+1; i++) {
            for (let j=1; j<m+1; j++) {
                if (graph[i][j] === req) graph[i][j] = ''
            }
        }
    }
    
    for (const request of requests) {
        if (request.length === 1) bfs(request)
        else two(request[0])
    }
    
    let ans = 0    
    for (let i=1; i<n+1; i++) {
        for (let j=1; j<m+1; j++) {
            if (graph[i][j] != '') ans++
        }
    }
    
    return ans;
}